clean_code_fm_guide.md

```diff
@@ -0,0 +1,410 @@
+
+**Bad:**
+```
+Mons.Game.exec_turn(hero_obj: Mons.Object, adve_obj: Mons.Object, ..., game: Mons.Game) : Mons.Game
+```
+
+**Good:**
+```
+Mons.Game.exec_turn(
+  hero_obj: Mons.Object, 
+  adve_obj: Mons.Object,
+  ...
+  game: Mons.Game) : Mons.Game
+```
+
+### Functions should do one thing
+This is by far the most important rule in software engineering. When functions do more than one thing, they are harder to compose, test, and reason about. When you can isolate a function to just one action, it can be refactored easily and your code will read much cleaner. If you take nothing else away from this guide other than this, you'll be ahead of many developers.
+
+**Bad:**
+```
+type Film {
+  new(title: String, year: Nat)
+}
+
+main: List(Film)
+  let films = [
+    Film.new("Fight Club ", 1999), 
+    Film.new("Avatar", 2009),
+    Film.new("Passengers", 2016),
+    Film.new("My Neighbor Totoro", 1988)
+    Film.new("The Silence of the Lambs", 1991)]
+  let wishlist = List.nil<Film>
+  for film in films with wishlist:
+    open film
+    if Nat.gtn(film.year, 2008) then
+      List.cons<_>(film, wishlist)
+    else 
+      wishlist
+
+```
+
+**Good:**
+```
+type Film {
+  new(title: String, year: Nat)
+}
+
+catalog: List(Film)
+  [ Film.new("Fight Club ", 1999), 
+    Film.new("Avatar", 2009),
+    Film.new("Passengers", 2016),
+    Film.new("My Neighbor Totoro", 1988)
+    Film.new("The Silence of the Lambs", 1991)]
+
+films_newer_than(year: Nat, films: List(Film)): List(Film)
+  List.filter<_>( 
+    (f) open f Nat.gtn(f.year, year), // anonymous function, that is, a function without a name
+    films)
+
+main: List(Film)
+  let wishlist = films_newer_than(2008, catalog)
+  wishlist
+```
+
+### Avoid negative conditionals
+**Bad:**
+```
+player_not_free_to_move(player: Player): Bool
+```
+
+**Good:**
+```
+is_player_free_to_move(player: Player): Bool
+```
+
+### Remove dead code 
+Dead code is just as bad as duplicate code. There's no reason to keep it in your codebase. If it's not being called, get rid of it! It will still be safe in your version history if you still need it.
+
+### Use auxiliary functions to get small information about your data
+```
+type Mons.Kind{
+  Mons(ele: Mons.Kind.mons, boss: Bool),
+  Terrain(ele: Mons.Kind.terrain),
+}
+
+type Mons.Kind.mons{
+  HERO,
+  MAGE,
+  POISOLICK,
+  EMERELDER
+}
+
+type Mons.Kind.terrain{
+  VOID,
+  FLOOR(lvl: U32, model: U32),
+  MON_AREA
+}
+
+Mons.Kind.is_hero(kind: Mons.Kind): Bool
+  case kind{
+    Mons: 
+    case kind.ele{
+      HERO     : true,
+      MAGE     : false,
+      POISOLICK: false,
+      EMERELDER: false,
+    }
+    Terrain: false
+  }
+
+Mons.Kind.is_mon_area(adve_kin: Mons.Kind): Bool
+  case adve_kin{
+    Mons: false
+    Terrain:
+    case adve_kin.ele{
+      VOID    : false
+      FLOOR   : false
+      MON_AREA: true
+    }
+  }
+
+```
+
+
+## **Data Structures**
+### Use getter and setter
```

Comment:
not sure I want to add that as a guideline because it should be a language feature, to derive get/set for you. forcing users to write them manually may feel laborious